app/src/main/java/fr/free/nrw/commons/nearby/NearbyPlaces.java

```diff
@@ -9,23 +10,156 @@
+                    "  (SAMPLE(COALESCE(?icon0, ?icon1)) as ?icon)\n" +
+                    "  (SAMPLE(COALESCE(?emoji0, ?emoji1)) as ?emoji)\n" +
+                    "WHERE {\n" +
+                    "  # Around given location...\n" +
+                    "  SERVICE wikibase:around {\n" +
+                    "    ?item wdt:P625 ?location.\n" +
+                    "    bd:serviceParam wikibase:center \"Point({{LONG}} {{LAT}})\"^^geo:wktLiteral. \n" +
+                    "    bd:serviceParam wikibase:radius \"{{RADIUS}}\" . # Radius in kilometers.\n" +
+                    "  }\n" +
+                    "  \n" +
+                    "  # ... and without an image.\n" +
+                    "  MINUS {?item wdt:P18 []}\n" +
+                    "  \n" +
+                    "  # Get the label in the preferred language of the user, or any other language if no label is available in that language.\n" +
+                    "  OPTIONAL {?item rdfs:label ?item_label_preferred_language. FILTER (lang(?item_label_preferred_language) = \"{{LANG}}\")}\n" +
+                    "  OPTIONAL {?item rdfs:label ?item_label_any_language}\n" +
+                    "  \n" +
+                    "  # Get the class label in the preferred language of the user, or any other language if no label is available in that language.\n" +
+                    "  OPTIONAL {\n" +
+                    "    ?item p:P31/ps:P31 ?classId.\n" +
+                    "    OPTIONAL {?classId rdfs:label ?class_label_preferred_language. FILTER (lang(?class_label_preferred_language) = \"{{LANG}}\")}\n" +
+                    "    OPTIONAL {?classId rdfs:label ?class_label_any_language}\n" +
+                    "\n" +
+                    "    # Get icon\n" +
+                    "    OPTIONAL { ?classId wdt:P2910 ?icon0. }\n" +
+                    "    OPTIONAL { ?classId wdt:P279*/wdt:P2910 ?icon1. }\n" +
+                    "    # Get emoji\n" +
+                    "    OPTIONAL { ?classId wdt:P487 ?emoji0. }\n" +
+                    "    OPTIONAL { ?classId wdt:P279*/wdt:P487 ?emoji1. }\n" +
+                    "  }\n" +
+                    "}\n" +
+                    "GROUP BY ?item\n";
+    private static List<Place> places;
+    private static NearbyPlaces singleton;
+    private double radius = RADIUS_INITIAL;
+
+    private NearbyPlaces(){
+    }
 
-    public static List<Place> get() {
-        if(places != null) {
-            return places;
+    public List<Place> getFromWikidataQuery(LatLng curLatLng, String lang) {
+        List<Place> places = Collections.emptyList();
+        while (true) {
+            places = getFromWikidataQuery(curLatLng, lang, radius);
+            if (places.size() >= MIN_RESULTS) {
+                break;
+            } else {
+                radius *= RADIUS_MULTIPLIER;
+            }
+        }
+        while (places.size() > MAX_RESULTS && radius > RADIUS_INITIAL) {
```

Comment:
Is there any drawback to getting "too many" places? If the places are sorted from nearest (or on a map) there should be no problem I guess?